### DISCUSSION TITLE
Question Regarding Yielded Vehicle Behaviour in Intersection Environment

### DISCUSSION THREAD

**Discussion by ehtnamuh** [eleurent/highway-env#469] [General]
I noticed when vehicles yield, they tend to shake and rotate randomly perhaps in an attempt to find a way out. Can someone tell me which part of the code causes this behaviour? I want to stop the cars from doing this but cannot find what is causing it. The recover_from_stop function never seems to get called so this behaviour must stem from somewhere else. Thank you.

**Comment by eleurent** [comment]
> I noticed when vehicles yield, they tend to shake and rotate randomly perhaps in an attempt to find a way out. Can someone tell me which part of the code causes this behaviour?

In principle they are not trying to find a way out, the lateral controller (which controls steering) is still tracking the same lane centerline, and the target speed of the longituinal (IDM) controller is set to 0. I also observed similar oscillations at low velocity in order circumstances, but if I remember correctly at the time I attributed it to rendering issues (i.e. the car's heading angle is not moving a lot, but the way pygame renders rotating rects made it shaky). It was long ago though, so its probably worth investigating again.

> The recover_from_stop function never seems to get called so this behaviour must stem from somewhere else. Thank you.

This function was introduced for another very specific usecase (vehicle stuck very close to an obstacle, and has to reverse before turning), and is disabled.

**Comment by ehtnamuh** [comment]
Thank you for your reply. The issue I was trying to describe actually entails the cars rotating to occupy half of an adjacent lane, which leads to collisions with other vehicles entering the lane. The collision detection doesn't seem to be coupled to rendering, but I can never be sure. The behaviour of reversing when the cars are stuck also seems to trigger if a car is stuck behind another car for an extended period of time. But I have no idea where this behaviour is stemming from in the code. Again thank you for your time and help. I'll see if I can figure out what is triggering the behaviour using breakpoints.

**Comment by eleurent** [comment]
> The issue I was trying to describe actually entails the cars rotating to occupy half of an adjacent lane, which leads to collisions with other vehicles entering the lane

Ah I see, sorry for the misunderstanding. I'm not really sure I have observed what you describe however, maybe its a recent regression. 

> The collision detection doesn't seem to be coupled to rendering, but I can never be sure

They are done independently, but in principle they should match.

> The behaviour of reversing when the cars are stuck also seems to trigger if a car is stuck behind another car for an extended period of time. But I have no idea where this behaviour is stemming from in the code

I don't think this behaviour is explicitly described in the code, but rather stems from the IDM model:

https://github.com/Farama-Foundation/HighwayEnv/blob/21225dbe12a50c088540c0257bcc05b03a97732e/highway_env/vehicle/behavior.py#L126

The vehicle's acceleration is computed from several variables, included the distance to its front vehicle, which is compared to a desired gap:

https://github.com/Farama-Foundation/HighwayEnv/blob/21225dbe12a50c088540c0257bcc05b03a97732e/highway_env/vehicle/behavior.py#LL158C9-L158C20

If the distance is lower than desired gap, this will result in negative acceleration. However, if the behaviour of reversing is undesirable, we may use this negative acceleration only for braking, and clip it to 0 when the speed reaches 0 or below.

**Comment by ehtnamuh** [comment]
Indeed the behaviour is caused by the IDM model. The issue with the cars turning in place when yielding for too long is caused by the steering controls. It seems the cars yield on a straight lane visually but, their path and their target lane is already a circular lane. Checking to see if the car is yielding and setting the steering to zero seems to solve the problem. Thank you for all your help.

**Reply by eleurent** [reply]
Nice investigation and solution, and thanks for letting me know! I'll probably make a fix
